Update diagram generation docs (#39)
anz-bank/sysl-website: docs/cmd-datamodel.md

sidebar_label: Data Model Diagram
---

:::info

> We are currently in the process of migrating from PlantUML to Mermaid for our diagram generation. This will remove the external dependency on PlantUML and offer a better user experience. Diagram generation with mermaid is currently supported for integration diagrams and sequence diagrams only. For more details, check out [sysl diagram](cmd-diagram)

Reviewer: Nit
```suggestion
We are currently in the process of migrating from PlantUML to Mermaid for our diagram generation. This will remove the external dependency on PlantUML and offer a better user experience. Diagram generation with Mermaid is currently supported for integration diagrams and sequence diagrams only. For more details, check out [sysl diagram](cmd-diagram).
```

:::

:::info

> This command requires the SYSL_PLANTUML environment variable to be set or passed in as a flag. Follow the instructions [here](plantuml.md) for more details

Reviewer: Nit
```suggestion
This command requires the `SYSL_PLANTUML` environment variable to be set or passed in as a flag. Follow the instructions [here](plantuml.md) for more details.
```

:::
---

> `sysl datamodel` lets you generate data model diagrams for types defined in Sysl.

Reviewer: Nit
```suggestion
`sysl datamodel generates data model diagrams for types defined in Sysl.
```

## Usage

```
usage: sysl datamodel [<flags>] <MODULE>
```

## Output Formats

> The output file format can be specified via the extension passed into the -o flag.
> Valid formats include .svg, .png, .uml, .puml, .plantuml, .html or .link

Reviewer: ```suggestion
The output file format can be specified via the extension of the filename passed into the `-o` flag.

Valid extensions include `.svg`, `.png`, `.uml`, `.puml`, `.plantuml`, `.html` or `.link`.
```

## Required Flags

> Either the project flag or the direct flag must be passed in. The direct flag generates data model diagrams for ALL types defined in the input sysl file. The project flag specifies a subset of applications to generate diagrams for. Refer to the [Project Datamodel Diagram](#project-datamodel-diagram) example for more info.

Reviewer: ```suggestion
Either the `project` flag or the `direct` flag must be passed in. The `direct` flag generates data model diagrams for ALL types defined in the input Sysl file. The `project` flag specifies a subset of applications to generate diagrams for. Refer to the [Project Datamodel Diagram](#project-datamodel-diagram) example for more info.
```

> - `-j, --project=PROJECT` Generate diagrams only for applications specified in the specified project

Reviewer: Nit
```suggestion
- `-j, --project=PROJECT` Generate diagrams only for applications listed in the specified project
```

- `-d, --direct` Generate diagrams for all applications and types

## Optional Flags

All flags are all optional.

Optional flags:

- `--class_format="%(classname)"`
- ` Specify the format string for data diagram participants. May include %%(appname) and %%(@foo) for attribute foo (default: %(classname))
- `-t, --title=TITLE` Diagram title
- `-p, --plantuml=PLANTUML` base url of PlantUML server (default: SYSL_PLANTUML or http://localhost:8080/plantuml see http://plantuml.com/server.html#install for more info)
- `-o, --output="%(epname).png"` Output file (default: %(epname).png)
- `-f, --filter=FILTER` Only generate diagrams whose names match a pattern

[More common optional flags](common-flags)
[Diagram format arguments](format-diagram)

## Arguments

Args:

- `<MODULE>` Input sysl file that contains the system specifications. e.g `simple.sysl`. The `.sysl` file type is optional.

## Examples

### Simple Datamodel Diagram

```bash

> sysl datamodel -d PaymentInfo.sysl -o PaymentInfo.svg

Reviewer: Nit: to match `help` output, keep input file at the end
```suggestion
sysl datamodel -d -o PaymentInfo.svg PaymentInfo.sysl
```

```

```sysl title="Input Sysl file: PaymentInfo.sysl"
Payment:
    !type CardInfo:
        credit_card_number <: string:
            @sensitive="true"
        type <: string
```

![Payment Datamodel diagram](/img/sysl/data-diagram-puml.svg)

### Compound Datamodel Diagram

```bash

> sysl datamodel -d Payment.sysl -o Payment.svg

Reviewer: ```suggestion
sysl datamodel -d -o Payment.svg Payment.sysl
```

```

```sysl title="Input Sysl file: Payment.sysl"
Payment:
    !type CardInfo:
        credit_card_number <: string:
            @sensitive="true"
        type <: string
    !type Payment:
        CardInfo <: CardInfo
        Amount <: int
```

![Payment Datamodel diagram](/img/sysl/data-diagram-compound-puml.svg)

### Project Datamodel Diagram

```bash

> sysl datamodel -j Project Payment.sysl -o "%(epname).svg"

Reviewer: ```suggestion
sysl datamodel -j Project -o "%(epname).svg" Payment.sysl
```

```

> In this example we generate two diagrams using the -j flag to specify which applications to display

Reviewer: ```suggestion
In this example we generate two diagrams using the `-j` flag to specify which applications to display.
```
